### Project Summary

The primary objective of the project is to apply the Capital Asset Pricing Model(CAPM) to analyse and determine the required compensation that an investor should expect given the risk to invest in an asset.The project begins by first estimating expected stock returns, followed by calculating the betas, determining the market return, and obtaining the risk-free rate. Finally, the project involves performing CAPM calculations and visualizing results to provide a comprehensive understanding of the stock's risk and investor expected compensation. Finally, culminating in a discussion of the results of this analysis. 

Concepts will be explained as the project progresses.

### What is CAPM?

The Capital Asset Pricing Model (CAPM) is a widely used financial model within the finance industry that helps in estimating the expected return on an investment based on its risk level. It considers the relationship between the asset's beta, which measures its volatility compared to the overall market, and the risk-free rate of return. By incorporating these factors, CAPM provides a framework for investors to assess the appropriate compensation they should receive for taking on the investment risk.

CAPM Formula: rf + (beta * (rm-rf))

### Notes on the project:

The dataset is time-series data extracted from Yahoo Finance named as "stocks_dataset.csv" with dates starting from 11/07/2013 and ending at 08/07/2020, the data is also based on the close price of the stocks. The data extracted is for Facebook(FB) - now Meta Platforms Inc (META), Twitter(TWTR), Boeing Co(BA), AT&T Inc(T), MGM Resorts International (MGM), Tesla (TSLA) and S&P500 Index. 

The rf or risk-free premium used in the CAPM Model is usually based on a 10 Year Treasury Bill Rate as it is widely accepted that the government will not default on their debts. I have obtained historical yield rates from the US treasury website matching the first date on my date column which was the 11/07/2013, this day had a rate of 3.31%.

The stock data will be normalized to make it easier to examine changes and trends in the data.

## Step 1: Importing the Libraries

In [1]:
from copy import copy
import numpy as np
import pandas as pd
import plotly.express as px

## Step 2: Importing the Dataset

In [2]:
df = pd.read_csv('/kaggle/input/capital-asset-pricing-model-capm-python-project/stocks_dataset.csv')

### Obtaining statistical information on the dataset

In [3]:
# This is just so we can gain a better understanding of the data

df.describe()

FB         TWTR         NFLX           BA            T  \
count  1699.000000  1699.000000  1699.000000  1699.000000  1699.000000   
mean    135.966863    31.296698   195.423654   218.325633    35.296986   
std      50.062473    11.508628   127.700712    98.968191     3.412142   
min      44.820000    14.010000    44.887142    95.010002    26.770000   
25%      88.080002    19.600001    90.474998   132.840004    32.790001   
50%     136.759995    31.610001   143.830002   158.289993    34.900002   
75%     178.075005    38.229999   320.024993   334.360001    38.004999   
max     268.440002    73.309998   548.729980   440.619995    43.470001   

               MGM         TSLA        sp500  
count  1699.000000  1699.000000  1699.000000  
mean     25.720865   311.299388  2413.219428  
std       5.200605   203.097452   424.825240  
min       7.140000   120.500000  1741.890015  
25%      21.935001   218.829994  2053.294921  
50%      25.879999   253.740005  2360.159912  
75%      29.250000   323.565002  2783.330078  
max      38.029999  1643.000000  3386.149902

## Step 3: Cleaning and Analysing the Dataset

### Creating a function to calculate the daily returns of the stock data

In [4]:
def daily_return(df):
    df_daily_return = df.copy()
    
    for i in df.columns[1:]:
        for j in range(1, len(df)):
            df_daily_return.loc[j, i] = ((df.loc[j, i] - df.loc[j-1, i]) / df.loc[j-1, i]) * 100
            
        df_daily_return.loc[0, i] = 0
    return df_daily_return

### Calculating the beta for stocks

In [5]:
# Note: beta represents the volatility of a stock in relation to the overall market movements. Where the market movement is equal
# to 1, stock beta that is greater than 1 suggests that a stock is more volatile than the market and a beta of less than 1 
# represents a stock with lower volatility than the market.

# For accessability I have opted to conduct each individual beta as oppose to a function so it is easier to understand what I
# am doing.

In [6]:
# Getting the daily returns for the stock

stock_daily_return = daily_return(df)

In [7]:
# Facebook

FB_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['FB'], 1)
print('Beta for {} stock is = {}'.format('FB', FB_Beta))

Beta for FB stock is = 1.096849088020326


In [8]:
# Twitter

TWTR_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['TWTR'], 1)
print('Beta for {} stock is = {}'.format('TWTR', TWTR_Beta))

Beta for TWTR stock is = 1.172359618315995


In [9]:
# Netflix

NFLX_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['NFLX'], 1)
print('Beta for {} stock is = {}'.format('NFLX', NFLX_Beta))

Beta for NFLX stock is = 1.0445724485449472


In [10]:
# Boeing Co

BA_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['BA'], 1)
print('Beta for {} stock is = {}'.format('BA', BA_Beta))

Beta for BA stock is = 1.4304608163084231


In [11]:
# AT&T Inc

T_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['T'], 1)
print('Beta for {} stock is = {}'.format('T', T_Beta))

Beta for T stock is = 0.7523861924715606


In [12]:
# MGM Resorts International

MGM_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['MGM'], 1)
print('Beta for {} stock is = {}'.format('MGM', MGM_Beta))

Beta for MGM stock is = 1.6517124147928917


In [13]:
# Tesla

TSLA_Beta, _ = np.polyfit(stock_daily_return['sp500'], stock_daily_return['TSLA'], 1)
print('Beta for {} stock is = {}'.format('TSLA', TSLA_Beta))

Beta for TSLA stock is = 1.2359723602006576


### Calculating the market rate of return (rm)

In [14]:
# Note: The NASDAQ is only open 252 days of the year so 252 days will be used instead of 365 days

rm = stock_daily_return['sp500'].mean() * 252

### Calculating the risk free rate (rf)

In [15]:
# As explained above the risk free premium or rf is based on the US 10 year treasury bill rate which is 3.31%. This number will
# need to be converted to a decimal to reflect the percentage value (3.31/100 = 0.0331)

rf = 3.31/100

## Calculating the expected returns using CAPM

In [16]:
# The CAPM formula is expressed as: rf + (beta * (rm-rf))

# Please note these results are in percentages (%).

# Risk-free premium = rf
# beta = stock beta
# rm = market rate of return

# For accessability I have opted to conduct each individual CAPM as oppose to a function so it is easier to understand what I
# am doing.


In [17]:
# Facebook Expected Return using CAPM Model

ER_FB = rf + (FB_Beta * (rm-rf))
print(ER_FB)

12.317208335243228


In [18]:
# Twitter Expected Return using CAPM Model

ER_TWTR = rf + (TWTR_Beta * (rm-rf))
print(ER_TWTR)

13.162884868810691


In [19]:
# Netflix Expected Return using CAPM Model

ER_NFLX = rf + (NFLX_Beta * (rm-rf))
print(ER_NFLX)

11.731738638699062


In [20]:
# Boeing Co. Expected Returns using CAPM Model

ER_BA = rf + (BA_Beta * (rm-rf))
print(ER_BA)

16.05347675809008


In [21]:
# AT&T Inc. Expected Returns using CAPM Model

ER_T = rf + (T_Beta * (rm-rf))
print(ER_T)

8.459412789249035


In [22]:
# MGM Resorts International. Expected Returns using CAPM Model

ER_MGM = rf + (MGM_Beta * (rm-rf))
print(ER_MGM)

18.531373339136046


In [23]:
# Tesla Expected Returns using CAPM Model

ER_TSLA = rf + (TSLA_Beta * (rm-rf))
print(ER_TSLA)

13.875312696255426


## Step 4: Visualizing the data (Interactive)

### Plotting the daily returns of stocks vs S&P 500 daily returns

In [24]:
# To gain a better understanding of the data, a visual aid would help with interpreting and understanding the data.

# By plotting these scatterplots we are able to visually examine and determine any potential correlation or relationship between 
# the two. Also I like looking at visuals rather than numbers and code.

In [25]:
for i in stock_daily_return.columns:
  
  if i != 'Date' and i != 'sp500':
    
    # Use plotly express to plot the scatter plot for every stock vs. the S&P500
    fig = px.scatter(stock_daily_return, x = 'sp500', y = i, title = i)

    # Fit a straight line to the data and obtain beta and alpha
    b, a = np.polyfit(stock_daily_return['sp500'], stock_daily_return[i], 1)
    
    # Plot the straight line 
    fig.add_scatter(x = stock_daily_return['sp500'], y = b*stock_daily_return['sp500'] + a)
    fig.show()

## Step 5: Discussing and Interpreting the results

The CAPM analysis yielded the following results:

Capital Asset Pricing Model Results:

Boeing: 16.05%
Tesla: 13.88%
Twitter: 13.16%
Facebook: 12.32%
Netflix: 11.73%
AT&T: 8.46%

These figures represent the estimated returns investors should expect based on the risk associated with each stock. Higher expected returns generally indicate higher levels of risk.

Among the stocks analyzed, Boeing stands out with the highest expected return of 16.05%, this suggests a relatively higher level of risk compared to the other stocks analysed. On the other hand, AT&T has the lowest expected return of 8.46%, indicating a relatively lower level of risk compared to the other stocks.

These findings provide insights into the potential returns of each stock during the specified time period, assisting investors in evaluating their investment strategies and allowing them to make more informed decisions.

CAPM should not be the only metric an investor relies on to invest. Instead, an investor should consider a confluence of metrics and financial analysis in order to gain better insight into their preferred assets.

## Step 6: Discussing Limitations

The analysis of CAPM results is subject to certain limitations that should be considered. 

Assumptions: CAPM relies on a lot of assumptions such as the assumption that markets are efficient, relationships are linear and betas are constant. 

Use of Historical Data: CAPM uses historical data to estimate risk and expected return however, past performance may not be indicative of future performance.

Data Quality and Availability: The accuracy and availability of the data is subjective as we have no idea how the original data was collected, cleaned, transformed or loaded.

Time-series data continuity: The data set only contains data from the last 7 years however, we have used a 10 year US Treasury Bill rate from 2013 which may not properly represent the bill rate or time-series data. 

## Project Conclusion

In conclusion, the Capital Asset Pricing Model (CAPM) is a valuable tool for assessing the expected return of an asset based on its systematic risk in relation to the overall market. By calculating the beta coefficient, which represents the asset's sensitivity to market movements, CAPM provides insights into the risk-reward tradeoff for investors. Our analysis of various stocks using CAPM revealed interesting levels of risk and expected returns. 

Stocks such as Boeing and Tesla exhibited higher betas, indicating greater volatility and risk with the potential for higher returns. On the other hand, AT&T displayed a lower beta, suggesting relatively lower risk and expected returns with lower risk. 

These findings can help investors make informed decisions in terms of their investment portfolios goals. However, it's important to note that CAPM relies on certain assumptions, such as efficient markets and the use of historical data, which may limit its accuracy in real-world scenarios. 

Nevertheless, CAPM is a valuable framework for evaluating investment opportunities and understanding the relationship between risk and expected returns.